File: lib/auth-service.ts
/**
 * Serviço de autenticação e validação de tokens
 */

import jwt from 'jsonwebtoken'
import { getPrismaClient, prismaPublic } from './prisma'

export interface TokenPayload {
  userId: string
  empresaId: string
  tenantId: string
  tenant: string
  role: string
  type: string
  iat: number
  exp: number
  iss: string
  aud: string
}

export interface ValidateTokenResult {
  valid: boolean
  payload?: TokenPayload
  error?: string
  usuario?: {
    id: string
    nome: string
    email: string
    role: string
    ativo: boolean
  }
  empresa?: {
    id: string
    razaoSocial: string
    nomeFantasia: string | null
    cnpj: string
    ativo: boolean
  }
  tenant?: {
    id: string
    subdomain: string
    active: boolean
  }
}

export class AuthService {
  /**
   * Valida um token JWT e retorna informações do usuário, empresa e tenant
   */
  static async validateToken(token: string): Promise<ValidateTokenResult> {
    try {
      const jwtSecret = process.env.JWT_SECRET || 'your-super-secret-jwt-key-change-in-production'

      // Verifica e decodifica o token
      const decoded = jwt.verify(token, jwtSecret, {
        issuer: 'solis-api',
        audience: 'solis-pdv',
      }) as TokenPayload

      // Busca o tenant no schema public
      const tenant = await prismaPublic.tenant.findUnique({
        where: { id: decoded.tenantId }
      })

      if (!tenant) {
        return {
          valid: false,
          error: 'Tenant não encontrado'
        }
      }

      // Valida se o tenant está ativo
      if (!tenant.active) {
        return {
          valid: false,
          error: 'Tenant está inativo'
        }
      }

      // Busca dados no schema do tenant
      const tenantPrisma = await getPrismaClient(decoded.tenant)
      
      // Busca o usuário
      const usuario = await tenantPrisma.usuario.findUnique({
        where: { id: decoded.userId }
      })

      if (!usuario) {
        return {
          valid: false,
          error: 'Usuário não encontrado'
        }
      }

      // Valida se o usuário está ativo
      if (!usuario.ativo) {
        return {
          valid: false,
          error: 'Usuário está inativo'
        }
      }

      // Busca a empresa
      const empresa = await tenantPrisma.empresa.findUnique({
        where: { id: decoded.empresaId }
      })

      if (!empresa) {
        return {
          valid: false,
          error: 'Empresa não encontrada'
        }
      }

      // Valida se a empresa está ativa
      if (!empresa.ativo) {
        return {
          valid: false,
          error: 'Empresa está inativa'
        }
      }

      return {
        valid: true,
        payload: decoded,
        usuario: {
          id: usuario.id,
          nome: usuario.nome,
          email: usuario.email,
          role: usuario.role,
          ativo: usuario.ativo
        },
        empresa: {
          id: empresa.id,
          razaoSocial: empresa.razaoSocial,
          nomeFantasia: empresa.nomeFantasia,
          cnpj: empresa.cnpj,
          ativo: empresa.ativo
        },
        tenant: {
          id: tenant.id,
          subdomain: tenant.subdomain,
          active: tenant.active
        }
      }

    } catch (error) {
      if (error instanceof jwt.TokenExpiredError) {
        return {
          valid: false,
          error: 'Token expirado'
        }
      }

      if (error instanceof jwt.JsonWebTokenError) {
        return {
          valid: false,
          error: 'Token inválido'
        }
      }

      return {
        valid: false,
        error: 'Erro ao validar token: ' + (error instanceof Error ? error.message : 'Erro desconhecido')
      }
    }
  }

  /**
   * Extrai informações básicas de um token sem validar no banco
   * Útil para debug ou logging
   */
  static decodeToken(token: string): TokenPayload | null {
    try {
      const decoded = jwt.decode(token) as TokenPayload
      return decoded
    } catch (error) {
      return null
    }
  }

  /**
   * Gera um token de autenticação para um usuário
   */
  static async generateToken(
    userId: string,
    empresaId: string,
    tenantId: string,
    expiresInDays: number = 30
  ): Promise<{ token: string; expiresAt: Date } | null> {
    try {
      // Valida o tenant
      const tenant = await prismaPublic.tenant.findUnique({
        where: { id: tenantId }
      })

      if (!tenant || !tenant.active) {
        return null
      }

      // Busca dados no schema do tenant
      const tenantPrisma = await getPrismaClient(tenant.subdomain)
      
      // Busca o usuário
      const usuario = await tenantPrisma.usuario.findUnique({
        where: { id: userId }
      })

      if (!usuario || !usuario.ativo) {
        return null
      }

      // Busca a empresa
      const empresa = await tenantPrisma.empresa.findUnique({
        where: { id: empresaId }
      })

      if (!empresa || !empresa.ativo) {
        return null
      }

      const jwtSecret = process.env.JWT_SECRET || 'your-super-secret-jwt-key-change-in-production'
      const expiresAt = new Date()
      expiresAt.setDate(expiresAt.getDate() + expiresInDays)

      const token = jwt.sign(
        {
          userId: usuario.id,
          empresaId: empresa.id,
          tenantId: tenant.id,
          tenant: tenant.subdomain,
          role: usuario.role,
          type: 'user',
          iat: Math.floor(Date.now() / 1000),
        },
        jwtSecret,
        {
          expiresIn: `${expiresInDays}d`,
          issuer: 'solis-api',
          audience: 'solis-pdv',
        }
      )

      return { token, expiresAt }
    } catch (error) {
      console.error('[AuthService] Erro ao gerar token:', error)
      return null
    }
  }

  /**
   * Gera um token de autenticação para agente PDV (sem usuário)
   */
  static async generateAgentToken(
    empresaId: string,
    tenantId: string,
    expiresInDays: number = 3650,
    agentName?: string
  ): Promise<{ token: string; expiresAt: Date } | null> {
    try {
      // Valida o tenant
      const tenant = await prismaPublic.tenant.findUnique({
        where: { id: tenantId }
      })

      if (!tenant || !tenant.active) {
        return null
      }

      // Busca a empresa no schema do tenant
      const tenantPrisma = await getPrismaClient(tenant.subdomain)
      const empresa = await tenantPrisma.empresa.findUnique({
        where: { id: empresaId }
      })

      if (!empresa || !empresa.ativo) {
        return null
      }

      const jwtSecret = process.env.JWT_SECRET || 'your-super-secret-jwt-key-change-in-production'
      const expiresAt = new Date()
      expiresAt.setDate(expiresAt.getDate() + expiresInDays)

      const token = jwt.sign(
        {
          userId: 'agent',
          empresaId: empresa.id,
          tenantId: tenant.id,
          tenant: tenant.subdomain,
          role: 'agent',
          type: 'agent',
          agentName: agentName || 'Agente PDV',
          iat: Math.floor(Date.now() / 1000),
        },
        jwtSecret,
        {
          expiresIn: `${expiresInDays}d`,
          issuer: 'solis-api',
          audience: 'solis-pdv',
        }
      )

      return { token, expiresAt }
    } catch (error) {
      console.error('[AuthService] Erro ao gerar token de agente:', error)
      return null
    }
  }

  /**
   * Autentica um usuário com email e senha
   */
  static async login(
    email: string,
    password: string,
    tenantSubdomain: string,
    empresaId: string
  ): Promise<{ success: boolean; token?: string; expiresAt?: Date; error?: string; usuario?: any }> {
    try {
      // Busca o tenant
      const tenant = await prismaPublic.tenant.findUnique({
        where: { subdomain: tenantSubdomain }
      })

      if (!tenant || !tenant.active) {
        return {
          success: false,
          error: 'Tenant não encontrado ou inativo'
        }
      }

      // Busca o usuário no schema do tenant
      const tenantPrisma = await getPrismaClient(tenantSubdomain)
      const usuario = await tenantPrisma.usuario.findUnique({
        where: { email: email.toLowerCase() }
      })

      if (!usuario) {
        return {
          success: false,
          error: 'Credenciais inválidas'
        }
      }

      // Verifica se o usuário está ativo
      if (!usuario.ativo) {
        return {
          success: false,
          error: 'Usuário inativo'
        }
      }

      // Verifica a senha (usando bcrypt)
      const bcrypt = require('bcryptjs')
      const passwordMatch = await bcrypt.compare(password, usuario.passwordHash)

      if (!passwordMatch) {
        return {
          success: false,
          error: 'Credenciais inválidas'
        }
      }

      // Verifica se a empresa existe e está ativa
      const empresa = await tenantPrisma.empresa.findUnique({
        where: { id: empresaId }
      })

      if (!empresa || !empresa.ativo) {
        return {
          success: false,
          error: 'Empresa não encontrada ou inativa'
        }
      }

      // Gera o token
      const tokenData = await AuthService.generateToken(
        usuario.id,
        empresaId,
        tenant.id,
        30 // 30 dias de validade
      )

      if (!tokenData) {
        return {
          success: false,
          error: 'Erro ao gerar token de autenticação'
        }
      }

      return {
        success: true,
        token: tokenData.token,
        expiresAt: tokenData.expiresAt,
        usuario: {
          id: usuario.id,
          nome: usuario.nome,
          email: usuario.email,
          role: usuario.role
        }
      }

    } catch (error) {
      console.error('[AuthService] Erro no login:', error)
      return {
        success: false,
        error: 'Erro ao realizar login'
      }
    }
  }
}

File: Services/AuthService.cs
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Microsoft.EntityFrameworkCore;
using Dapper;
using Npgsql;
using SolisApi.Data;
using SolisApi.DTOs;
using BCrypt.Net;

namespace SolisApi.Services;

/// <summary>
/// DTO interno para query de usuário
/// </summary>
internal class UserDto
{
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Email { get; set; } = string.Empty;
    public string Password_Hash { get; set; } = string.Empty;
    public string Role { get; set; } = string.Empty;
    public bool Active { get; set; }
    public DateTime Created_At { get; set; }
    public DateTime Updated_At { get; set; }
}

/// <summary>
/// DTO interno para query de empresa
/// </summary>
internal class CompanyDto
{
    public Guid Id { get; set; }
}

/// <summary>
/// Serviço de autenticação - geração e validação de tokens JWT
/// </summary>
public class AuthService
{
    private readonly string _jwtSecret;
    private readonly string _jwtIssuer;
    private readonly string _jwtAudience;
    private readonly SolisDbContext _context;
    private readonly IConfiguration _configuration;
    private readonly ILogger<AuthService> _logger;

    public AuthService(
        IConfiguration configuration, 
        SolisDbContext context,
        ILogger<AuthService> logger)
    {
        _jwtSecret = configuration["Jwt:Secret"] ?? throw new InvalidOperationException("JWT Secret não configurado");
        _jwtIssuer = configuration["Jwt:Issuer"] ?? "SolisApi";
        _jwtAudience = configuration["Jwt:Audience"] ?? "SolisApi";
        _context = context;
        _configuration = configuration;
        _logger = logger;
    }

    private string GetConnectionString(string tenantSubdomain)
    {
        var baseConnectionString = _configuration.GetConnectionString("DefaultConnection");
        var builder = new NpgsqlConnectionStringBuilder(baseConnectionString)
        {
            SearchPath = $"tenant_{tenantSubdomain},public"
        };
        return builder.ToString();
    }

    /// <summary>
    /// Gerar token JWT para usuário (30 dias de validade)
    /// </summary>
    public string GenerateToken(TokenPayload payload)
    {
        return GenerateTokenWithExpiry(payload, TimeSpan.FromDays(30));
    }

    /// <summary>
    /// Gerar token JWT para agente (10 anos de validade)
    /// </summary>
    public string GenerateAgentToken(TokenPayload payload)
    {
        payload.Type = "agent";
        return GenerateTokenWithExpiry(payload, TimeSpan.FromDays(3650)); // 10 anos
    }

    /// <summary>
    /// Gerar token com tempo de expiração customizado
    /// </summary>
    private string GenerateTokenWithExpiry(TokenPayload payload, TimeSpan expiry)
    {
        var tokenHandler = new JwtSecurityTokenHandler();
        var key = Encoding.UTF8.GetBytes(_jwtSecret);

        var claims = new List<Claim>
        {
            new("userId", payload.UserId.ToString()),
            new("empresaId", payload.EmpresaId.ToString()),
            new("tenantId", payload.TenantId.ToString()),
            new("tenant", payload.Tenant),
            new("role", payload.Role),
            new("type", payload.Type)
        };

        var tokenDescriptor = new SecurityTokenDescriptor
        {
            Subject = new ClaimsIdentity(claims),
            Expires = DateTime.UtcNow.Add(expiry),
            Issuer = _jwtIssuer,
            Audience = _jwtAudience,
            SigningCredentials = new SigningCredentials(
                new SymmetricSecurityKey(key),
                SecurityAlgorithms.HmacSha256Signature)
        };

        var token = tokenHandler.CreateToken(tokenDescriptor);
        return tokenHandler.WriteToken(token);
    }

    /// <summary>
    /// Validar token JWT e retornar o payload
    /// </summary>
    public TokenPayload? ValidateToken(string token)
    {
        try
        {
            var tokenHandler = new JwtSecurityTokenHandler();
            var key = Encoding.UTF8.GetBytes(_jwtSecret);

            var validationParameters = new TokenValidationParameters
            {
                ValidateIssuerSigningKey = true,
                IssuerSigningKey = new SymmetricSecurityKey(key),
                ValidateIssuer = true,
                ValidIssuer = _jwtIssuer,
                ValidateAudience = true,
                ValidAudience = _jwtAudience,
                ValidateLifetime = true,
                ClockSkew = TimeSpan.Zero
            };

            var principal = tokenHandler.ValidateToken(token, validationParameters, out _);

            var userId = principal.FindFirst("userId")?.Value;
            var empresaId = principal.FindFirst("empresaId")?.Value;
            var tenantId = principal.FindFirst("tenantId")?.Value;
            var tenant = principal.FindFirst("tenant")?.Value;
            var role = principal.FindFirst(ClaimTypes.Role)?.Value; // Use ClaimTypes.Role instead of "role"
            var type = principal.FindFirst("type")?.Value;

            return new TokenPayload
            {
                UserId = Guid.Parse(userId ?? string.Empty),
                EmpresaId = Guid.Parse(empresaId ?? string.Empty),
                TenantId = Guid.Parse(tenantId ?? string.Empty),
                Tenant = tenant ?? string.Empty,
                Role = role ?? string.Empty,
                Type = type ?? string.Empty
            };
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Falha ao validar token JWT");
            return null;
        }
    }

    /// <summary>
    /// Login - validar credenciais e retornar token
    /// </summary>
    public async Task<LoginResponse?> LoginAsync(string tenantSubdomain, LoginRequest request)
    {
        _logger.LogInformation("Tentativa de login para tenant {Tenant}, email {Email}", tenantSubdomain, request.Email);

        // Buscar tenant
        var tenant = await _context.Tenants
            .FirstOrDefaultAsync(t => t.Subdomain == tenantSubdomain && t.Active);

        if (tenant == null)
        {
            _logger.LogWarning("Tenant {Tenant} não encontrado ou inativo", tenantSubdomain);
            return null;
        }

        // Conectar ao schema do tenant usando Dapper
        using var connection = new NpgsqlConnection(GetConnectionString(tenantSubdomain));
        await connection.OpenAsync();

        // Buscar usuário por email
        var user = await connection.QueryFirstOrDefaultAsync<UserDto>(@"
            SELECT id, name, email, password_hash, role, active, created_at, updated_at
            FROM users
            WHERE LOWER(email) = LOWER(@Email) AND active = true
        ", new { request.Email });

        if (user == null)
        {
            _logger.LogWarning("Usuário {Email} não encontrado ou inativo no tenant {Tenant}", request.Email, tenantSubdomain);
            return null;
        }

        // Verificar senha
        if (!BCrypt.Net.BCrypt.Verify(request.Password, user.Password_Hash))
        {
            _logger.LogWarning("Senha inválida para usuário {Email} no tenant {Tenant}", request.Email, tenantSubdomain);
            return null;
        }

        // Buscar primeira empresa ativa (default)
        var company = await connection.QueryFirstOrDefaultAsync<CompanyDto>(@"
            SELECT id
            FROM companies
            WHERE active = true
            LIMIT 1
        ");

        if (company == null)
        {
            _logger.LogWarning("Nenhuma empresa ativa encontrada no tenant {Tenant}", tenantSubdomain);
            return null;
        }

        // Gerar token
        var payload = new TokenPayload
        {
            UserId = user.Id,
            EmpresaId = company.Id,
            TenantId = tenant.Id,
            Tenant = tenant.Subdomain,
            Role = user.Role,
            Type = "user"
        };

        var token = GenerateToken(payload);

        _logger.LogInformation("Login bem-sucedido para usuário {UserId} ({Email}) no tenant {Tenant}", 
            user.Id, user.Email, tenantSubdomain);

        return new LoginResponse
        {
            Token = token,
            User = new UserResponse
            {
                Id = user.Id,
                Name = user.Name,
                Email = user.Email,
                Role = user.Role,
                Active = user.Active,
                CreatedAt = user.Created_At,
                UpdatedAt = user.Updated_At
            }
        };
    }
}
